8259822: [PPC64] Support the prefixed instruction format added in POWER10

kazunoriogata opened the pull request:
The POWER10 processor, which implements Power ISA 3.1 [1], supports new instruction formats where an instruction takes two 32bit words.  The first word is called prefix, and the instructions with prefix are called prefixed instructions.  With more bits in opcode and operand fields, POWER10 supports larger immediate value in an operand, as well as many new instructions.

This is the first changes to handle prefixed instructions, and this adds support of prefixed addi (= paddi) instruction as an example of prefix usage.  paddi accepts 34bit immediate value, while original addi accepts 16bit value.

[1] https://ibm.ent.box.com/s/hhjfw0x0lrbtyzmiaffnbxh2fuo0fog0

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259822](https://bugs.openjdk.java.net/browse/JDK-8259822): [PPC64] Support the prefixed instruction format added in POWER10


### Reviewers
 * [Corey Ashford](https://openjdk.java.net/census#cashford) (@CoreyAshford - Author)
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2095/head:pull/2095` \
`$ git checkout pull/2095`

Update a local copy of the PR: \
`$ git checkout pull/2095` \
`$ git pull https://git.openjdk.java.net/jdk pull/2095/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2095`

View PR using the GUI difftool: \
`$ git pr show -t 2095`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2095.diff">https://git.openjdk.java.net/jdk/pull/2095.diff</a>

</details>

kazunoriogata commented:
I ran jtreg test on POWER10 box by using "make test-tier1" and verified no additional fails.

kazunoriogata commented:
/issue add JDK-8259822

CoreyAshford commented on src/hotspot/cpu/ppc/assembler_ppc.hpp:
```diff
@@ -1424,6 +1515,32 @@ class Assembler : public AbstractAssembler {
+  static int is_paddi(const int* p, bool is_pli = false) {
+     int32_t* p_inst = (int32_t*)p;
+
+     if (is_aligned(reinterpret_cast<uintptr_t>(p_inst+1), 64) && is_nop(*p_inst)) {
```
This test is a bit confusing.  Shouldn't is_paddi return false if p points at a nop (even if it precedes the paddi)?

kazunoriogata replied:
"is_<instr>()" are the functions for testing instructions in the generated code and they are usually called in "assert()" or complex combination of if statements.  So if this function returns false when p points at nop, the assertions or the if-conditions need to handle the nop case and the code will become difficult to read.  Since the padding nop is a non-common case, I'd like to hide the existence of nop in this function.

CoreyAshford replied:
Ok, got it.  Thanks for the explanation.  Maybe an extra comment in the code saying essentially what you said here would be appropriate?

CoreyAshford replied:
Looks good!

CoreyAshford commented on src/hotspot/cpu/ppc/assembler_ppc.hpp:
```diff
@@ -1424,6 +1515,32 @@ class Assembler : public AbstractAssembler {
+        return false;
+     }
+  }
+  static bool is_pli(const int* p)   { return is_paddi(p, true); }
```
is_pli() is defined, but not used, as far as I can tell.

kazunoriogata replied:
It is just intended to show a usage of the second argument of is_paddi(const int* p, bool is_pli = false).  Is it unnecessary, i.e., self-evident?

CoreyAshford replied:
In past open source projects I have worked on, functions that essentially are unreachable in the final product cannot be tested, so there's no way to know if they work as intended. For that reason, only code that is used and/or can be tested in the final product (e.g. via an externally-visible API) would be accepted in a commit.

I don't know if the OpenJDK project follows that rule or not.

CoreyAshford commented on src/hotspot/cpu/ppc/ppc.ad:
```diff
@@ -5887,6 +5919,21 @@ instruct loadConL32_Ex(iRegLdst dst, immL32 src) %{
+// Load 34-bit long constant using prefixed addi. No constant pool entries required.
+instruct loadConL34(iRegLdst dst, immL34 src) %{
+  match(Set dst src);
+  ins_cost(DEFAULT_COST+1);
```
There's no predicate for >= POWER10.  I can see how this works because of the immL34 operand having its own predicate, but in later instructs, e.g. addL_reg_imm34 even though the operand is immI32, you still add the explicit predicate.

I'd rather there be an explicit POWER10 predicate in this instruct.

kazunoriogata replied:
If predicate is added, adlc fails with an error message: "Syntax Error: :ADLC does not support instruction chain rules with predicates"  I think addL_reg_imm34 allows predicate because it is not called from other rules.  Is it better to leave some comments?  (BTW, immI32 is only for POWER10 or higher.  POWER9 version uses immI16 or immI16hi.)

CoreyAshford replied:
Hmm, I'm confused.  I don't see any other reference to loadConL34 in ppc.ad.

CoreyAshford replied:
I wish we knew why this predicate is causing an issue, but I guess it's not important because the operand type provides sufficient limiting of the instruct.

CoreyAshford commented on src/hotspot/cpu/ppc/assembler_ppc.inline.hpp:
```diff
@@ -168,6 +200,20 @@ inline void Assembler::lis(  Register d, int si16)             { Assembler::addi
+inline void Assembler::pla(  Register d, Register a, long si34) { Assembler::paddi( d, a, si34, false); }
+inline void Assembler::psubi(Register d, Register a, long si34) { Assembler::paddi( d, a, -si34, false); }
+
+inline void Assembler::pli_or_li(Register d, long si34) {
```
Defined but not used.

kazunoriogata replied:
It is intended to show an example of an efficient set of macros/functions for a prefixed instruction.  Since a non-prefixed (32-bit) instruction is more efficient if the argument fits within the operand fields because of smaller I-cache usage and no need for the padding nop, it is convenient to have a macro/function that select better format.  I intended to show an example of recommended set of useful macros/functions when adding support for a prefixed instruction.  I'm planning to use this function in my next patch.  Is it better to move this to the next patch?

kazunoriogata replied:
Same reason as the next comment (paddi_or_addi), but necessity of this one may be weaker.

CoreyAshford replied:
Ok.  See my other comments about unused/untestable code.

CoreyAshford commented on src/hotspot/cpu/ppc/assembler_ppc.inline.hpp:
```diff
@@ -130,6 +143,25 @@ inline void Assembler::divw_(  Register d, Register a, Register b) { emit_int32(
+  emit_int32( PADDI_SUFFIX_OPCODE | rt(d)   | ra(a)   | d1_eo(si34));
+}
+
+inline void Assembler::paddi_or_addi(Register d, Register a, long si34) {
```
defined but not used.

kazunoriogata replied:
It is intended to show an example of an efficient set of macros/functions for a prefixed instruction. Since a non-prefixed (32-bit) instruction is more efficient if the argument fits within the operand fields because of smaller I-cache usage and no need for the padding nop, it is convenient to have a macro/function that select better format. I intended to show an example of recommended set of useful macros/functions when adding support for a prefixed instruction. I'm planning to use this function in my next patch. Is it better to move this to the next patch?

CoreyAshford replied:
It would be my preference to post code up only when it's actually used.  When you get around to using it, there's a possibility you will want to modify it a bit, add a comment, etc.   So that can be done all in one commit, instead of split across two.

CoreyAshford commented on src/hotspot/cpu/ppc/assembler_ppc.cpp:
```diff
@@ -50,6 +55,22 @@ int AbstractAssembler::code_fill_byte() {
+#ifdef COMPILER2
+  Thread* thread = Thread::current();
+
+  if(!thread->is_Compiler_thread())       return false;
```
add space after "if"

CoreyAshford reviewed (commented):
This looks good overall.  I'm looking forward to being able to utilize this capability.

kazunoriogata commented on src/hotspot/cpu/ppc/assembler_ppc.hpp:
```diff
@@ -1081,6 +1109,20 @@ class Assembler : public AbstractAssembler {
   static int inv_bo_field(int x)  { return inv_opp_u_field(x, 10,  6); }
   static int inv_bi_field(int x)  { return inv_opp_u_field(x, 15, 11); }
 
+  // support to extended opcodes (prefixed instructions) introduced by POWER10
```
I'll leave it since other comments that refer to a specific ISA implementation use Power X. (I changed POWER10 to Power 10, though.)

CoreyAshford replied:
ok, sounds good.

kazunoriogata commented on src/hotspot/cpu/ppc/ppc.ad:
```diff
@@ -5816,6 +5834,20 @@ instruct loadConI32_lo16(iRegIdst dst, iRegIsrc src1, immI16 src2) %{
+            "PLI     $dst, $src" %}
+  size(12);
+  ins_encode %{
+    // TODO: PPC port $archOpcode(ppc64Opcode_paddi);
```
This comment was simply copied from old code that was deleted in JDK-8253029.  I removed the commend in this patch, too.

kazunoriogata commented:
@CoreyAshford Thank you very much for your review.  I updated my patch based on your comment.  For some comments that I left unchanged, I replied to each of them.  Please give me your thought.

For merge conflict, I verified the conflict is only the copyright years.  I'll fix it when real changes are reviewed because pulling the latest master to resolve conflict makes this patch difficult to review.

kazunoriogata commented:
@CoreyAshford Thank you for your comment.  Regarding the unused code, I agree that unused code won't be tested enough, so I removed paddi_or_addi(), pli_or_li(), and is_pli().

For the comment on predicate in loadConI32 and loadConL34, I couldn't find the reason why it caused build error.  So I added comments describing they are only for Power 10 and up but can't add predicate.  I also added comments in is_paddi() as you suggested.

CoreyAshford commented on src/hotspot/cpu/ppc/assembler_ppc.inline.hpp:
```diff
@@ -34,6 +34,19 @@ inline void Assembler::emit_int32(int x) {
+
+  // Add nop if a prefixed (two-word) instruction is going to cross a 64-byte boundary.
+  // (See Section 1.6 of Power ISA Version 3.1)
+  if(is_aligned(reinterpret_cast<uintptr_t>(pc()) + sizeof(int32_t), 64) ||
```
add space after 'if'

CoreyAshford commented on src/hotspot/cpu/ppc/assembler_ppc.inline.hpp:
```diff
@@ -130,6 +143,17 @@ inline void Assembler::divw_(  Register d, Register a, Register b) { emit_int32(
+// Prefixed instructions, introduced by POWER10
+inline void Assembler::paddi(  Register d, Register a, long si34, bool r = false) {
+  assert(a != R0 || r, "r0 not allowed, unless R is set (CIA relative)");
+  paddi_r0ok( d, a, si34, r);
```
The space after the ( isn't needed here, since it's not aligning with a similar call above or below.

CoreyAshford commented on src/hotspot/cpu/ppc/assembler_ppc.hpp:
```diff
@@ -1245,6 +1305,22 @@ class Assembler : public AbstractAssembler {
+    return (((int *)a)[instruction_number] << 14) >> 14;
+  }
+
+  static inline int hi18_signed(  int x) { return hi16_signed(x); }
+  static inline int hi18_signed( long x) { return (int)((x << 30) >> 46); }
```
The extra spaces for alignment look a bit strange here.  I think it should be:
```
static inline int hi18_signed( int x) ...
static inline int hi18_signed(long x) ...
```
Basically, remove the leading space from the ( long x) one and delete one from (  int x) to match.

CoreyAshford commented on src/hotspot/cpu/ppc/assembler_ppc.hpp:
```diff
@@ -1301,9 +1380,18 @@ class Assembler : public AbstractAssembler {
+  // Prefixed add immediate, introduced by POWER10
+  inline void paddi(Register d, Register a, long si34, bool r);
+  inline void pli(  Register d, long si34);
+  inline void pla(  Register d, long si34);
+  inline void pla(  Register d, Register a, long si34);
+  inline void psubi(Register d, Register a, long si34);
```
pla and psubi are defined but not used.

CoreyAshford reviewed (changes requested):
Just a few more minor things... several defined-but-not-used functions, and some little formatting issues.

kazunoriogata commented:
@CoreyAshford Thank you for your review.  I went through my  changes to check spacing, as well as removed unused pla and psubi as you pointed out.

CoreyAshford reviewed (approved):
Looks good!  Thanks for the additional formatting cleanup

TheRealMDoerr commented:
Sorry that I didn't review it, yet. I didn't have enough time.

Please note that C2 already has a mechanism to handle alignment:

Insert "ins_alignment(n);" into the instruct. This allows C2 to insert up to n-1 nop instructions. 
Add a "compute_padding" function to determine the actual number of nops to insert given the current offset.
They are used in s390.ad. Please take a look.

kazunoriogata commented:
@CoreyAshford Thank you for your review.

@TheRealMDoerr Thank you for your suggestion.  I think I understand the mechanism.  I'll update my patch after verifing my change and running jtreg.

kazunoriogata commented:
@TheRealMDoerr  Sorry to be late to update the patch.  It took time to debug my careless mistakes...

I updated the patch with using ins_alignment and verified there is no additional errors in jtreg tier1 tests.  Although there were intermittent failures in VarHandleTests using weakCompareAndSet, it was the same situation as the base build, so I think it is not related to my changes.

TheRealMDoerr reviewed (changes requested):
Thanks for contributing this and for using ins_alignment. Update needs further changes.

kazunoriogata commented:
@TheRealMDoerr  Thank you for your review.  I updated the patch based on your comments.  As a result of this, all changes in assembler_ppc.cpp were reverted.

```src/hotspot/cpu/ppc/ppc.ad
@@ -5912,7 +6020,7 @@ instruct loadConL(iRegLdst dst, immL src, iRegLdst toc) %{
   ins_field_cbuf_insts_offset(int);
   format %{ "LD      $dst, offset, $toc \t// load long $src from TOC" %}
 - size(4);
 + size(12);
```

I picked up a diff for other changes by my mistake.  Thank you for pointing this out.

```src/hotspot/cpu/ppc/assembler_ppc.hpp
  static int is_paddi(const int* p, bool is_pli = false) {
     int32_t* p_inst = (int32_t*)p;

     if (is_aligned(reinterpret_cast<uintptr_t>(p_inst+1), 64) && is_nop(*p_inst)) {
```
> TheRealMDoerr
> Do we still need to skip over the nop? I think this should no longer happen.

I think it is still useful when JIT scans a sequence of generated code like `MacroAssembler::get_const()`.  Assuming a generated sequence:

```label:
  paddi rA, const32a(r0)
  nop
  paddi rB, const32b(r0)
```

it would be convenient if `get_const()` can check the next instruction after the first `paddi` without taking care of the nop, and it can start scanning from the nest instruction address (= label+8).

TheRealMDoerr commented on src/hotspot/cpu/ppc/assembler_ppc.hpp:
```diff
@@ -1424,6 +1507,28 @@ class Assembler : public AbstractAssembler {
+     } else {
+        return false;
+     }
+  }
```
I can't see any usage of it. As Corey stated, unused code should better get removed because it doesn't get tested and may break unless you're planning to use it in the near future or need it for debugging purposes.

TheRealMDoerr reviewed (changes requested):
Thanks for the update. Looks basically good to me, but the merge conflict needs to get resolved. Also, please check for unused code.

kazunoriogata commented:
@TheRealMDoerr  Thank you for your review.

>  TheRealMDoerr
> I can't see any usage of it. As Corey stated, unused code should better get removed because it doesn't get tested and may break unless you're planning to use it in the near future or need it for debugging purposes.

I see.  These code were locally used for supporting `paddi` in `load_const()`/`get_const()`/`patch_const()`, so I intended to use it soon.  However, it still has some errors.  So I'll move these changes to the patch after fixing the errors.

> TheRealMDoerr
> The Copyright year change in this file was already done by another change. You need to revert it in order to resolve the merge conflict.

Thank you for the tip!  I intended to merge to the latest code base when all other changes are reviewed, but reverting the lines is much easier to get the same result.

CoreyAshford reviewed (changes requested):
No functional comments.  Just some minor grammar things.

kazunoriogata commented:
@CoreyAshford Thank you for your comment.  (I want an automatic grammar checker in github...)

> Since this string is used so many times, I almost feel like it ought to be factored out.

I factored out the function bodies of `*Node::compute_padding()` to a static function, since they are just duplicated.

TheRealMDoerr commented on src/hotspot/cpu/ppc/ppc.ad:
```diff
@@ -5887,6 +5977,24 @@ instruct loadConL32_Ex(iRegLdst dst, immL32 src) %{
+instruct loadConL34(iRegLdst dst, immL34 src) %{
+  match(Set dst src);
+  // This macro is valid only in Power 10 and up, but adding the following predicate here
+  // caused build error. So we comment it out for now.
```
> 
> 
> // caused **a** build error **, so** we comment it out for now.

Better would be, "so we only check it in the operand".

TheRealMDoerr reviewed (approved):
Looks good to me, now. Is the latest version substantially tested? We need to rely on IBMs testing because nobody else has Power10.

kazunoriogata commented:
Thank you for a heads-up.  I've just run jtreg teir1 tests with the latest version.  I'll run DaCapo, Renaissance, and SPECjbb2015 to verify the latest version.

kazunoriogata commented:
@TheRealMDoerr  I verified there was no additional errors against the base version in DaCapo, Renaissance, and SPECjbb2015.  Altough some tests in DaCapo and Renaissance (e.g., DaCapo eclipse and tradesoap and Renaissance db-shootout) failed in the modified version, they also failed in the base build with the same error.

kazunoriogata commented:
/integrate

kazunoriogata commented:
@TheRealMDoerr  @CoreyAshford  Thank you for your review.  I think this is now ready to be integrated.

@TheRealMDoerr  Could you sponsor this pull request?

TheRealMDoerr commented:
Yes, I can sponsor it. Thanks for testing.
@CoreyAshford Are all your requests resolved? Do you need more time for additional tests? Please approve once you're fine with it.

CoreyAshford commented:
The only other thing I see is that there are a number of static functions that are defined for accessing fields (e.g. `inv_st_x1`), but aren't actually used yet.  However, I think these are probably ok, since they are not very complex.  So I will approve.

TheRealMDoerr commented:
/sponsor

kazunoriogata commented:
@TheRealMDoerr  Thank you for sponsoring this patch.
